### ISSUE TITLE
nix cannot build more than 1012 derivations in a single shot

### ISSUE THREAD

**Issue by ryantrinkle** [NixOS/nix#875]
This is a narrow example, but the problem arose in a real-world situation, when compressing many files for an asset pipeline.  Each file is processed as a separate derivation to ensure that changing a single file doesn't cause other files to be re-compressed.  Overall, this system seems to work well.

However, with large file counts - specifically, anything over 1012 derivations - this fails.  For example:

``` bash
nix-build --option use-binary-caches false -E 'with import <nixpkgs> {}; map (x: stdenv.mkDerivation { name = builtins.toString x; builder = builtins.toFile "builder.sh" "echo $x > $out"; inherit x; }) (lib.range 1 1013)'
```

This code produces `error: writing to file: Broken pipe`.  However, with the `1013` changed to `1012`, it works.  Other situations seem to exhaust the number of file handles available.

Note that derivations that are already built do not contribute to this count - so if you run successfully with 1012 derivations, you'll need to increase the number to 2025 to see the error again.

**Comment by Ericson2314** [comment]
Sounds like its running out of file descriptors?

**Comment by ryantrinkle** [comment]
@Ericson2314 Yes, that's at least one of the limitations here.  On my machine, nix-daemon appears to be limited to 4096 or so file descriptors, and it creates 4 per derivation.

**Comment by copumpkin** [comment]
Seems like it would be fairly easy to make it self-limit based on how many
FDs it has available

**Comment by ryantrinkle** [comment]
@copumpkin: That makes sense, although I wonder if it even needs to take out so many locks in the first place.  Here's where I think it's taking the locks: https://github.com/NixOS/nix/blob/8cffec84859cec8b610a2a22ab0c4d462a9351ff/src/libstore/build.cc#L1287

This appears to me to be locking everything that will be built; however, isn't it sufficient to only lock what is actually being worked on?  If two nix-builds are building the same set of derivations, they can interleave freely without damage, afaik, provided they don't work on the same derivation at the same time.

**Comment by shlevy** [comment]
@ryantrinkle That line only takes locks for the current derivation being built :smile:

**Comment by ryantrinkle** [comment]
@shlevy Hmm, the behavior to me has appeared to be that nix-build won't start building *anything* until it can acquire locks on *everything*.  So, suppose you have build A depends on C and D, and build B depends on D and E.  If A is building and has not yet finished C or D, I think the builder for B won't start on E until A has finished with D.

My suggestion was that it may not be necessary to lock derivations that aren't being actively worked on - so, for example, if each builder only builds one derivation at a time, A's builder could lock only C, then only D, then only A.  Of course, that might mean that another builder "steals" one of these builds, but I'm not sure there would be a problem with that.

**Comment by shlevy** [comment]
I think at least that the behavior you're asking for is intended by nix, though it may be buggy and actually waiting to obtain *all* the locks.

**Comment by domenkozar** [comment]
refs https://github.com/NixOS/nix/pull/3522
